boot/freeldr/freeldr/freeldr.spec

```diff
@@ -1,47 +1,184 @@
+@ stdcall -arch=i386 ScsiPortWritePortUshort(ptr long)
+@ stdcall -arch=i386 ScsiPortWritePortUlong(ptr long)
+@ stdcall -arch=i386 ScsiPortWriteRegisterBufferUchar(ptr ptr long)
+@ stdcall -arch=i386 ScsiPortWriteRegisterBufferUshort(ptr ptr long)
+@ stdcall -arch=i386 ScsiPortWriteRegisterBufferUlong(ptr ptr long)
+@ stdcall -arch=i386 ScsiPortWriteRegisterUchar(ptr long)
+@ stdcall -arch=i386 ScsiPortWriteRegisterUshort(ptr long)
+@ stdcall -arch=i386 ScsiPortWriteRegisterUlong(ptr long)
+
+# ARC
+@ cdecl ArcClose()
+@ cdecl ArcGetDiskCount()
+@ cdecl ArcGetDiskInfo()
+@ cdecl ArcGetFileInformation()
+@ cdecl ArcGetTime()
+@ cdecl ArcOpen()
+@ cdecl ArcRead()
+@ cdecl ArcSeek()
+
+# Debug
+@ cdecl DbgParseDebugChannels()
+@ cdecl DbgPrint(str)
+@ cdecl DbgPrint2(long long str long str)
+@ cdecl DebugDumpBuffer()
+@ cdecl DebugInit()
+@ cdecl FrLdrBugCheckWithMessage()
+@ stdcall KeBugCheckEx(long long long long long)
+
+# Heap
+@ cdecl FrLdrHeapAllocateEx()
+@ cdecl FrLdrHeapFreeEx()
+@ cdecl FrLdrHeapAlloc()
+@ cdecl FrLdrHeapFree()
+@ cdecl FrLdrTempAlloc()
+@ cdecl FrLdrTempFree()
+@ cdecl FrLdrHeapCleanupAll()
+
+# INI (check if we can move this to rosload)
+@ cdecl IniAddSection()
+@ cdecl IniAddSettingValueToSection()
+@ cdecl IniCleanup()
+@ cdecl IniGetFileSectionListHead()
+@ cdecl IniGetNumSectionItems()
+@ cdecl IniGetSectionSettingNameSize()
+@ cdecl IniGetSectionSettingValueSize()
+@ cdecl IniFileInitialize()
+@ cdecl IniModifySettingValue()
+@ cdecl IniOpenSection()
+@ cdecl IniReadSettingByName()
+@ cdecl IniReadSettingByNumber()
+
+# Mm
+@ cdecl AddMemoryDescriptor()
+@ cdecl MmAllocateHighestMemoryBelowAddress()
+@ cdecl MmAllocateMemoryAtAddress()
+@ cdecl MmAllocateMemoryWithType()
+@ cdecl MmFreeMemory()
+@ cdecl MmGetBiosMemoryMap()
+@ cdecl MmGetHighestPhysicalPage()
+@ cdecl MmGetLoaderPagesSpanned()
+@ cdecl MmGetMemoryMap()
+@ cdecl MmGetSystemMemoryMapTypeString()
+@ cdecl MmGetTotalPagesInLookupTable()
+
+# NtLdr options
+@ cdecl NtLdrGetNextOption()
+@ cdecl NtLdrGetOption()
+@ cdecl NtLdrGetOptionEx()
+@ cdecl NtLdrGetOptionExN()
+@ cdecl NtLdrAddOptions()
+
+# PeLdr
+@ cdecl PeLdrAllocateDataTableEntry()
+@ cdecl PeLdrCheckForLoadedDll()
+@ cdecl PeLdrFreeDataTableEntry()
+@ cdecl PeLdrImportDllLoadCallback()
+@ cdecl PeLdrInitSecurityCookie()
+@ cdecl PeLdrLoadImage()
+@ cdecl PeLdrLoadImageEx()
+@ cdecl PeLdrScanImportDescriptorTable()
+
+# UI
+@ cdecl UiDisplayMenu()
+@ cdecl UiDrawBackdrop()
+@ cdecl UiDrawProgressBarCenter()
+@ cdecl UiDrawStatusText()
+@ cdecl UiDrawText()
+@ cdecl UiEditBox()
+@ cdecl UiGetMenuBgColor()
+@ cdecl UiGetScreenHeight()
+@ cdecl UiIndicateProgress()
+@ cdecl UiInitialize()
+@ cdecl UiMessageBox()
+@ cdecl UiMessageBoxCritical()
+@ cdecl UiResetForSOS()
+@ cdecl UiSetProgressBarSubset()
+@ cdecl UiSetProgressBarText()
+@ cdecl UiShowMessageBoxesInArgv()
+@ cdecl UiShowMessageBoxesInSection()
+@ cdecl UiUnInitialize()
+@ cdecl UiUpdateProgressBar()
+@ cdecl TuiPrintf()
```

Comment:
For the UI, it could be of interest to export a (pointer to) the Ui functions table, `UiVtbl`
Maybe the same could be done for the Mach* functions? (`MachVtbl`)

Reply:
I'm not a fan of exporting data. It makes it neccessary to compile everything seperately with different _declspec(dllimport) definitions. As soon as you put something in a library things break.